torch/fx/experimental/symbolic_shapes.py

```diff
@@ -4234,31 +4238,32 @@ def _smart_symbol_sort(x):
                 floor_div_atoms = lhs.atoms(FloorDiv).union(rhs.atoms(FloorDiv))
                 if len(floor_div_atoms) > 0 and any(a.divisor != 1 for a in floor_div_atoms):
                     raise NotImplementedError
-                # short-circuit when no solving is needed
-
-                if isinstance(lhs, sympy.Symbol) and free_unbacked_symbols(lhs):
+                # Never replace unbacked symbols with other unbacked symbols.
+                # This is error prone because you can cause references to
+                # unbacked symbols to time travel backwards.  E.g.,
+                #
+                # u1 = x.item()
+                # ... use of u1 ...
+                # u2 = y.item()
+                # u3 = z.item()
+                # torch._check(u1 == u2 + u3)
+                #
+                # If you replace u1 with u2 + u3, then the use of u1 now
+                # references u2 and u3 prior to them actually being bound at
+                # runtime.  It's pretty inconvenient to setup control
+                # dependencies for substitutions, so ban it entirely.
```

Comment:
One way to do this is to declare that we just start using a substitution if all the unbacked symbols in it are defined. This also has the neat property that it "floats" the substitution as high up as they can go, meaning that
```
                # u1 = x.item()
                # ... use of u1 ...
                # u2 = y.item()
                # u3 = z.item()
                # .,.. use of u2, u3 ...
                # torch._check(u1 == u2 + u3)
```
becomes the example you wrote above

Reply:
While it might be possible to apply extra substitutions when we are evaluating guards, in general simplification can occur in unpredictable points whenever you access the .expr field of a SymNode. Also, applying extra substitutions when evaluating guards is also tricky, because we do multiple passes through the graph (e.g., when retracing) and whenever we start a new pass, we need to "reset" the set of unbacked symbols that are defined.

If we are relaxing this, I think I much prefer "unbacked SymInts have a topological order and you can only replace into earlier ordered SymInts." But I think this case is also difficult to do 100% correctly. Imagine you have:

```
y: f32[u0] = f(x)
z: f32[u1] = g(y)
```

We do a decomposition on g which introduces a fresh unbacked symbol:

```
y: f32[u0] = f(x)
z0: f32[u2] = g1(y)
z: f32[u1] = g2(z0)
```

Now the topological order of unbacked SymInts is not the same as a the order they were allocated. I need to know to insert u2 between u0 and u1 in the ordering, and I'm not even sure this maintenance is even easy to do in some of the optimization contexts.

This example is a bit contrived, but it's difficult enough that I don't really want to do anything about it unless we actually need it.

Reply:
I guess I'll do something like, only unbacked SymInts that occurred on initial traversal of the graph have an order, and then anything that shows up later is banned from participating in replacements. Which I don't expect the latter case to ever occur.

Reply:
>we do multiple passes through the graph (e.g., when retracing) and whenever we start a new pass, we need to "reset" the set of unbacked symbols that are defined.

Sure, but we now have sources for symints, so the replacements could be done in terms of these sources, which should be pass-agnostic.

> nly unbacked SymInts that occurred on initial traversal of the graph have an order, and then anything that shows up later is banned from participating in replacements

this is also a reasonable approach

Reply:
You mean the keypaths. I don't see how this helps though. The keypath only makes sense relative to an FX node that produces the thing you run the keypath getter on, and you don't necessarily have an FX node when you're doing simplification.